### PR TITLE
Screen stream settings

### PR DESCRIPTION
**Pull request by Diadlo**
Added ability of one screen and screen region selection
And refactor of avform (should I squash it in one commit?)

spectacle wz1410

Need for Windows and OSX tests (I think, that region selection tests will be enough, because multi screen use same principle)

### CODE REVIEW

**Comment by zetok** [comment]
#1786, #2548, #3303

**Comment by zetok** [comment]
on c2212c4:
- [x] when sharing a tiny fraction of the screen there was some strange video output
  [2016-06-15 qtox desktop sharing video output.webm.zip](https://github.com/tux3/qTox/files/316297/2016-06-15.qtox.desktop.sharing.video.output.webm.zip)

**Comment by Diadlo** [comment]
@zetok it's not a part of showing video?

**Comment by zetok** [comment]
That green/blue/red stuff shouldn't be there.

Also, from stdout:

```
[15:19:55.296] src/video/netcamview.cpp:88 : Debug: SIZER QRect(0,0 1920x1080)
```

**Comment by Diadlo** [comment]
@zetok Fixed :)

**Comment by zetok** [comment]
on 468a796:
- [x] selecting screen region is done every time `Audio / Video` tab in settings is being opened, even though there is already set region
- [x] during video call, when switched to `A / V` tab in settings I've tried to not pick any region, by pressing `ESC`, and trying to close window. Resulted in:
  
  ```
  14:20:51.130] src/widget/form/settings/avform.cpp:175 : Debug: opened!
  [14:20:51.138] src/video/videoframe.cpp:130 : Warning: None of the frames are valid! Did someone release us?
  [14:20:51.140] src/video/videoframe.cpp:130 : Warning: None of the frames are valid! Did someone release us?
  [x11grab @ 0x555557912be0] XShmGetImage() failed
  [x11grab @ 0x555557912be0] XShmGetImage() failed
  ```
  
  Even though I did end up selecting region, both camera preview and side receiving video only see black image. `[x11grab @ 0x555557912be0] XShmGetImage() failed` is continuously printed to stdout on the sending side.
  
  Changing region during video call without unnecessarily pressing other keys seemed to work just fine.
  
  Outside of video call doing the same results in:
  
  ```
  [14:27:42.903] src/widget/tool/screenshotgrabber.cpp:212 : Debug: Rejected screenshot
  [x11grab @ 0x555557c95180] XShmGetImage() failed
  ```
- [x] there was some strange message in stdout, should all these values be all `0` ?
  
  ```
  14:19:01.948] src/video/netcamview.cpp:88 : Debug: SIZER QRect(0,0 0x0)
  ```

**Comment by zetok** [comment]
on 468a796:
- [x] after ending the call after stuff mentioned above, memleak: 
  memleak

File: src/video/cameradevice.cpp [hunk 1]
```diff
@@ -55,7 +55,10 @@ CameraDevice* CameraDevice::open(QString devName, AVDictionary** options)
     CameraDevice* dev = openDevices.value(devName);
     int aduration;
     if (dev)
-        goto out;
```
Comment (sudden6) [inline src/video/cameradevice.cpp:58]: why remove the goto here?

Comment (Diadlo) [inline src/video/cameradevice.cpp:58]: I don't want start holy war about "goto". Many people absolutely against goto statement[1]. Why don't abstain if we can?

[1] [Edgar Dijkstra: Go To Statement Considered Harmful ](https://files.ifi.uzh.ch/rerg/arvo/courses/kvse/uebungen/Dijkstra_Goto.pdf)

Comment (sudden6) [inline src/video/cameradevice.cpp:58]: I agree about not using goto in general, but IMHO this is the exact case where it's usefull.

Comment (Diadlo) [inline src/video/cameradevice.cpp:58]: Make 1 line instead of 2? Don't think, that it's really useful.

File: src/video/cameradevice.h [hunk 2]
```diff
@@ -69,11 +66,15 @@ class CameraDevice
     /// or the system default.
     static QString getDefaultDeviceName();
 
+    /// Checks, that device name specify the display
```
Comment (sudden6) [inline src/video/cameradevice.h]: I think that should mean: `Checks, if a device name specifies a display`

File: src/video/cameradevice.cpp (hunk 1 above)
Comment (sudden6) [inline src/video/cameradevice.cpp:58]: No, in this case the goto ensures that in every case where the function is left the lock is unlocked, and removing all the gotos also causes a lot of code duplication.

Both introduce unnecessary possibilites for new mistakes, which could be avoided by that goto.

Comment (Diadlo) [inline src/video/cameradevice.cpp:58]: You think, that better revert this part?

File: src/widget/form/settings/avform.cpp [hunk 3]
```diff
@@ -136,56 +138,70 @@ void AVForm::showEvent(QShowEvent* event)
...
+    if (devIndex < 0 || devIndex >= videoDeviceList.size())
     {
         qWarning() << "Invalid device index";
         return;
     }
     QString devName = videoDeviceList[devIndex].first;
     VideoMode mode = videoModes[index];
-    Settings::getInstance().setCamVideoRes(QSize(mode.width, mode.height));
+
+    if (CameraDevice::isScreen(devName) && !mode.height && !mode.width) {
```
Comment (sudden6) [inline src/widget/form/settings/avform.cpp]: opening `{` on new line^^

File: src/widget/form/settings/avform.cpp [hunk 4]
```diff
@@ -205,109 +221,153 @@ void AVForm::updateVideoModes(int curIndex)
...
         bodyUI->videoModescomboBox->addItem(str);
     }
+
     if (videoModes.isEmpty())
         bodyUI->videoModescomboBox->addItem(tr("Default resolution"));
+
+    bodyUI->videoModescomboBox->blockSignals(previouslyBlocked);
+}
+
+int AVForm::searchPreveredIndex()
```
Comment (sudden6) [inline src/widget/form/settings/avform.cpp]: `searchPreferred Index()´

File: src/widget/form/settings/avform.cpp [hunk 5]
```diff
@@ -205,109 +221,153 @@ void AVForm::updateVideoModes(int curIndex)
...
+    }
+
+    int preferedIndex = searchPreveredIndex();
+    if (preferedIndex!= -1)
+    {
+        bodyUI->videoModescomboBox->setCurrentIndex(preferedIndex);
+        return;
     }
+
+    // If the user hasn't set a preffered resolution yet,
```
Comment (sudden6) [inline src/widget/form/settings/avform.cpp]: also preferred

File: src/video/cameradevice.h (hunk 2 above)
Comment (zetok) [inline src/video/cameradevice.h]: without the comma

File: src/video/cameradevice.cpp (hunk 1 above)
Comment (sudden6) [inline src/video/cameradevice.cpp:58]: yep

File: src/widget/form/settings/avform.cpp (hunk 4 above)
Comment (Diadlo) [inline src/widget/form/settings/avform.cpp]: I think I was very tired)

**Comment by Diadlo** [comment]
@zetok [First list](https://github.com/tux3/qTox/pull/3392#issuecomment-226769008) seems fixed

**Comment by zetok** [comment]
on bcca305f93acea4d7a0798038d1eccf83462f58d:
- [x] region setting is not being kept across qTox restarts – after restart / logout setting is being reset to `screen 1`, setting reset can happen even without restart, although only region would be shown in video preview
- [x] memleak is still there memleak stronk

**Comment by Diadlo** [comment]
@zetok You think, that we should restore region after restart/change device?

**Comment by zetok** [comment]
> @zetok You think, that we should restore region after restart/change device?

Well, I would expect video setting to be set to the same region after stopping qTox and starting it again.

Doing it differently could result in users accidentally streaming their whole desktop, which could be bad.

**Comment by Diadlo** [comment]
@zetok Fixed

**Comment by zetok** [comment]
on b0c716e32468e99490e0872ab7d56da3c24a1188:
- [x] region selection is kept only when logging out/in, still not kept across restarts

edit:
- [x] memleak still stronk – 14GB memleaked during 7min video call
  dat 14G

**Comment by Diadlo** [comment]
@zetok Sorry. I pushed old code with `-f`. Try now, please. It should be normal

**Comment by zetok** [comment]
on 3c736ebae3e7739545ba97075c430a3099f70efd:
- [x] memleak leaks more slowly
- [x] after restart region setting still doesn't seem to be preserved – preview windows shows whole screen, even though in stdout:
  
  ```
  [21:39:17.410] src/video/netcamview.cpp:88 : Debug: SIZER QRect(1645,1 18x18)
  ```
- [ ] also some bitrate thing repeated a few times?
  
  ```
  [21:39:30.001] src/core/coreav.cpp:611 : Debug: Recommended bitrate with 10  is now  64 / 6080 , ignoring it
  ```

**Comment by Diadlo** [comment]
@zetok 
1) Fixed
2) Can't reproduce on 2 devices. Can you describe steps?
3) Hm.. I don't know what is it

**Comment by sudden6** [comment]
@Diadlo 
3) Is probably not from your changes

**Comment by zetok** [comment]
@Diadlo 
1. set region
2. turn off qTox, and turn it on
3. start video call – whole screen is used

**Comment by Diadlo** [comment]
@zetok Fixed

**Comment by zetok** [comment]
cc @RowenStipe @towlie 
Could you perhaps test how well it works on osx & windows?

**Comment by towlie** [comment]
It doesn't compile for windows:

```
../src/platform/camera/directshow.cpp: In function 'QVector<VideoMode> DirectShow::getDeviceModes(QString)':
../src/platform/camera/directshow.cpp:237:1: error: jump to label 'nextformat' [-fpermissive]
 nextformat:
 ^
../src/platform/camera/directshow.cpp:228:26: note:   from here
                     goto nextformat;
                          ^
../src/platform/camera/directshow.cpp:230:27: note:   crosses initialization of 'VideoMode mode'
                 VideoMode mode;
                           ^
../src/platform/camera/directshow.cpp:237:1: error: jump to label 'nextformat' [-fpermissive]
 nextformat:
 ^
../src/platform/camera/directshow.cpp:224:26: note:   from here
                     goto nextformat;
                          ^
../src/platform/camera/directshow.cpp:230:27: note:   crosses initialization of 'VideoMode mode'
                 VideoMode mode;
                           ^
Makefile.Release:11764: recipe for target 'release/directshow.o' failed
make[1]: *** [release/directshow.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/fedora/Desktop/qtox/qTox-0d2a96b738cfa03e3032777424e3a18733e976cd/build'
Makefile:34: recipe for target 'release' failed
make: *** [release] Error 2

```

**Comment by Diadlo** [comment]
@towlie Can you test it now?

File: src/platform/camera/directshow.cpp [hunk 6]
```diff
@@ -214,20 +215,21 @@ QVector<VideoMode> DirectShow::getDeviceModes(QString devName)
...
-            for (int i=0; i<n; ++i)
+            for (int i = 0; i < n; ++i)
             {
                 AM_MEDIA_TYPE* type = nullptr;
+                VideoMode mode;
                 if (config->GetStreamCaps(i, &type, (BYTE*)vcaps) != S_OK)
                     goto nextformat;
 
                 if (!IsEqualGUID(type->formattype, FORMAT_VideoInfo)
                     && !IsEqualGUID(type->formattype, FORMAT_VideoInfo2))
```
Comment (Diadlo) [inline src/platform/camera/directshow.cpp]: This two conditions can be true in one time?

**Comment by towlie** [comment]
https://gist.github.com/towlie/758cb381da8b1ee853778c03152cebda

**Comment by Diadlo** [comment]
@towlie Fixed

**Comment by towlie** [comment]
Everything seems to work now. (Windows)

**Comment by towlie** [comment]
Oh and here is a log: https://gist.github.com/towlie/8ad4c394edf46ab908de9c48aa4c29f0

@zetok is this error already known? (looks like an issue which has nothing todo with this PR)

```
[15:32:31.164] ../src/main.cpp:169 : Debug: Log file over 1MB, rotating...
[15:32:31.165] ../src/main.cpp:177 : Warning: Unable to remove old log file
[15:32:31.172] ../src/main.cpp:180 : Critical: Unable to move logs
```

**Comment by zetok** [comment]
@towlie nope

**Comment by Diadlo** [comment]
I think #2018 related from this

**Comment by zetok** [comment]
@qTox/core anyone time for merge?

**Comment by zetok** [comment]
@antis81 @sudden6 time for merge?

**Comment by PKEv** [comment]
Cool!!!

**Pull request merged**